ray-operator/controllers/ray/rayjob_controller_scheduled_test.go

```diff
@@ -0,0 +1,211 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ray
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+	"github.com/ray-project/kuberay/ray-operator/controllers/ray/common"
+	"github.com/ray-project/kuberay/ray-operator/controllers/ray/utils"
+)
+
+var _ = Context("RayJob with schedule operation", func() {
+	Describe("When creating a RayJob with a schedule field and NO cluster deletion", Ordered, func() {
+		// The states should transition from Scheduled -> ... -> Initializing  ->  Running -> Complete -> Scheduled
+		// In the last scheduled state the cluster should still exist since ShutdownAfterJobFinishes is False
+		ctx := context.Background()
+		namespace := "default"
+		cronSchedule := "0 0 1 1 *"
+		rayJob := rayJobTemplate("rayjob-scheduled-no-deletion", namespace)
+		rayJob.Spec.Schedule = cronSchedule
+		rayJob.Spec.ShutdownAfterJobFinishes = false
+		rayCluster := &rayv1.RayCluster{}
+
+		It("Verify RayJob spec", func() {
+			Expect(rayJob.Spec.ShutdownAfterJobFinishes).To(BeFalse())
+			Expect(rayJob.Spec.Schedule).To(Not(BeEmpty()))
+		})
+
+		It("should create a RayJob object with the schedule", func() {
+			err := k8sClient.Create(ctx, rayJob)
+			Expect(err).NotTo(HaveOccurred(), "failed to create test scheduled RayJob resource")
+		})
+
+		It("should start in the Scheduled state", func() {
+			Eventually(
+				getRayJobDeploymentStatus(ctx, rayJob),
+				time.Second*60, time.Microsecond*500).Should(Equal(rayv1.JobDeploymentStatusScheduled),
+				"JobDeploymentStatus should be Initializing")
+		})
+
+		It("should NOT create a raycluster object while scheduled", func() {
+			Consistently(
+				getRayClusterNameForRayJob(ctx, rayJob),
+				time.Second*3, time.Millisecond*500).Should(BeEmpty())
+		})
+
+		It("should update schedule string in its spec", func() {
+			rayJobLookupKey := types.NamespacedName{Name: rayJob.Name, Namespace: rayJob.Namespace}
+			fetchedRayJob := &rayv1.RayJob{}
+			newSchedule := "*/1 * * * *"
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, rayJobLookupKey, fetchedRayJob)
+				return err == nil
+			}, time.Second*5, time.Millisecond*200).Should(BeTrue(), "expected to fetch the RayJob before update")
+
+			fetchedRayJob.Spec.Schedule = newSchedule
+
+			err := updateRayJobScheduleField(ctx, fetchedRayJob, newSchedule)
+			Expect(err).NotTo(HaveOccurred(), "failed to update RayJob's schedule in spec")
+
+			Eventually(func() bool {
```

Comment:
Please follow this pattern to keep the test code easier to read https://github.com/ray-project/kuberay/blob/master/ray-operator/controllers/ray/rayjob_controller_test.go#L217-L219

Reply:
I think this Eventually block is redundant, no need to test that the object has the updated field if there was no error returned on line 83. It's safe to assume the API is working as intended